packages/components/src/internal/components/project/ProjectSettings.tsx

```diff
@@ -10,38 +10,102 @@ import { resolveErrorMessage } from '../../util/messaging';
+import { ProjectDataTypeSelections } from './ProjectDataTypeSelections';
 import { DeleteProjectModal } from './DeleteProjectModal';
 
 export interface ProjectSettingsProps {
-    onChange: () => void;
+    onChange: (dirty?: boolean) => void;
     onPageError: (e: string) => void;
-    onSuccess: () => void;
+    onSuccess: (dirty?: boolean, reload?: boolean, isDelete?: boolean) => void;
+    project?: Container;
 }
 
-export const ProjectSettings: FC<ProjectSettingsProps> = memo(({ onChange, onSuccess, onPageError }) => {
+export const ProjectSettings: FC<ProjectSettingsProps> = memo(props => {
+    const { onChange, onSuccess, onPageError, project } = props;
+
+    const [loaded, setLoaded] = useState<boolean>(false);
+    const [disabledTypesMap, setDisabledTypesMap] = useState<{ [key: string]: number[] }>(undefined);
     const [showModal, setShowModal] = useState<boolean>(false);
-    const [dirty, setDirty] = useState<boolean>(false);
+    const [nameDirty, setNameDirty] = useState<boolean>(false);
+    const [dataTypeDirty, setDataTypeDirty] = useState<boolean>(false);
+    const [storageDirty, setStorageDirty] = useState<boolean>(false);
+    const [barDirty, setBarDirty] = useState<boolean>(false);
+    const [nameIdDirty, setNameIdDirty] = useState<boolean>(false);
     const [error, setError] = useState<string>();
     const [isSaving, setIsSaving] = useState<boolean>(false);
     const { api } = useAppContext();
-    const { container, user } = useServerContext();
+    const { projectDataTypes, ProjectFreezerSelectionComponent } = useAdminAppContext();
+    const { container, user, moduleContext } = useServerContext();
     const dispatch = useServerContextDispatch();
 
-    const onChange_ = useCallback(() => {
-        setDirty(true);
-        onChange();
+    useEffect(() => {
+        (async () => {
+            setLoaded(false);
+            setError(undefined);
+
+            try {
+                const disabledTypesMap_ = await getFolderDataTypeExclusions(project.path);
+                setDisabledTypesMap(disabledTypesMap_);
+            } catch (e) {
+                setError(`Error: ${resolveErrorMessage(e)}`);
+            } finally {
+                setLoaded(true);
+            }
+        })();
+    }, [project]);
+
+    const onNameChange_ = useCallback(() => {
+        setNameDirty(true);
+        onChange(true);
     }, [onChange]);
 
-    const onSubmit = useCallback(
+    const onDataTypeChange_ = useCallback(() => {
+        setDataTypeDirty(true);
+        onChange(true);
+    }, [onChange]);
+
+    const onStorageChange_ = useCallback(() => {
+        setStorageDirty(true);
+        onChange(true);
+    }, [onChange]);
+
+    const onBarChange_ = useCallback(() => {
+        setBarDirty(true);
+        onChange(true);
+    }, [onChange]);
+
+    const onBarSuccess_ = useCallback(() => {
+        setBarDirty(false);
+        onSuccess(nameDirty || dataTypeDirty || storageDirty || nameIdDirty, false);
```

Comment:
this check happens a lot. I'd suggest adding a isDirty variable that uses `useMemo` or something like that so there is just a single OR statement like this. That way if a new prop like this is added, it is easier to update.

Reply:
Unfortunately this doesn't work... I tried it out and state resulted in a weird state, because the overallDirty variable is dependent on the individual dirty bit that's being altered. For example, onBarSuccess_ alters barDirty, but anyDirty is calculated using barDirty (and others)

Reply:
ah yes, that makes sense. bummer